### DISCUSSION TITLE
Is the models' object.objects.bulk_update() and bulk_create() function supported?

### DISCUSSION THREAD

**Discussion by jchambers2012** [netbox-community/netbox#16972] [Q&A]
We have a few scripts that do large bulk operations like synchronizing a hundred VLANs to interfaces with tags or pull in IPAM data from external sources. Today we use the `object.save()` takes some time to run (like .3 sec per vlan per interface – one run took 60+ minutes).  We wanted to start using the function `bulk_update()` to help speedup this operation based on some limited testing in our development environment.  Reading the docs for bulk_update the `update()` function is called and the `pre_save` and `post_save` signals aren’t sent.  Will this be an issue if we convert our scripts to use bulk_update vs a loop of saves?

https://docs.djangoproject.com/en/5.0/ref/models/querysets/#bulk-update
> Each model’s `save()` method isn’t called, and the `pre_save` and `post_save` signals aren’t sent.

https://docs.djangoproject.com/en/5.0/ref/models/querysets/#bulk-create
> The model’s `save()` method will not be called, and the `pre_save` and `post_save` signals will not be sent.

**Comment by jchambers2012** [comment]
We do see bulk_update is use in the code base - few examples:
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/ipam/signals.py#L16
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/ipam/signals.py#L26
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/core/models/data.py#L216
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/dcim/models/devices.py#L1263
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/extras/models/customfields.py#L275

And bulk_create is used in the testing files - few examples:
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/netbox/tests/test_search.py#L44
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/virtualization/tests/test_api.py#L163

- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/ipam/tests/test_views.py#L409
- https://github.com/netbox-community/netbox/blob/36df9228a61f2d2a400b5a511641e291af2c7cee/netbox/circuits/tests/test_filtersets.py#L466

**Comment by candlerb** [comment]
> Reading the docs for bulk_update the `update()` function is called and the `pre_save` and `post_save` signals aren’t sent. Will this be an issue if we convert our scripts to use bulk_update vs a loop of saves?

It can be.

Example: I wrote a custom script which used bulk_create() to add missing components to devices (from the device type). However, because the hooks aren't called, it didn't update the cached counters of the number of each type of component on a device, which meant that the tabs in the GUI either showed the wrong count, or the tabs were missing completely (if the count == 0)

Actually the [pull request](https://github.com/netbox-community/customizations/pull/101) which fixes that hasn't been merged.  It changes:

```
                    for i in items:
                        i.full_clean()
                    klass.objects.bulk_create(items)
```

to:

```
                    for i in items:
                        i.full_clean()
                        i.save()
```

**Comment by jchambers2012** [comment]
That what I was thinking as I was reading the source code more. It seems like it might an unsafe operation, might have to submit a feature request to add support for update() in models that support custom save().

Guessing we will have to go model-by-model to see what one might be safe and what might not be...... or just live with 60+ minute run time for some of our deployment scripts. 😄
